managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java

```diff
@@ -2363,13 +2363,13 @@ private void maybeOffload(CompletableFuture<PositionImpl> finalPromise) {
                                     + ", total size = {}, already offloaded = {}, to offload = {}",
                             name, toOffload.stream().map(LedgerInfo::getLedgerId).collect(Collectors.toList()),
                             sizeSummed, alreadyOffloadedSize, toOffloadSize);
+                    offloadLoop(unlockingPromise, toOffload, PositionImpl.LATEST, Optional.empty());
                 } else {
                     // offloadLoop will complete immediately with an empty list to offload
                     log.debug("[{}] Nothing to offload, total size = {}, already offloaded = {}, threshold = {}",
                             name, sizeSummed, alreadyOffloadedSize, threshold);
+                    unlockingPromise.complete(PositionImpl.LATEST);
                 }
-
-                offloadLoop(unlockingPromise, toOffload, PositionImpl.LATEST, Optional.empty());
```

Comment:
Is the change necessary? After this change, `offloadLoop` won't be completed exceptionally if the `toOffload` is empty when the managed ledger is closed.

Reply:
If the managed ledger has been closed, and the toOffload is empty, we just complete the future has no bad effect.

Reply:
I see. But the code here is very weird. It looks like the result of the future has no effect because `maybeOffload` accepts `NULL_OFFLOAD_PROMISE`, which is a completed future.

```java
            unlockingPromise.whenComplete((res, ex) -> {
                    offloadMutex.unlock();
                    // The following code is meaningless because `finalPromise` is already completed
                    if (ex != null) {
                        finalPromise.completeExceptionally(ex);
                    } else {
                        finalPromise.complete(res);
                    }
                });
```

And I've checked the calls of `maybeOffload`, it is only called in `maybeOffloadInBackground`, which could be called by:
1. `ledgerClosed`: the future is `NULL_OFFLOAD_PROMISE`.
2. `maybeOffload` when locking the `offloadMutex` successfully: since `maybeOffload` could only be called in `maybeOffloadInBackground`, the future must be `NULL_OFFLOAD_PROMISE` as well.

The code design here is really confused.

Reply:
Yes, the behavior is the same, but only make it more readable.

Reply:
But it has the comments:

> // offloadLoop will complete immediately with an empty list to offload